### PR TITLE
[FEATURE][ML] Client-side API methods for DataFrame analytics config.

### PR DESCRIPTION
**Pull request by przemekwitek**
This PR implements basic lifecycle methods (Get/Put/Delete) for DataFrame analytics configs in high-level REST client.

### CODE REVIEW

File: client/rest-high-level/src/main/java/org/elasticsearch/client/MachineLearningClient.java [hunk 1]
```diff
@@ -1877,4 +1882,129 @@ public void setUpgradeModeAsync(SetUpgradeModeRequest request, RequestOptions op
             listener,
             Collections.emptySet());
     }
+
+    /**
+     * Create a new data frame analytics config
+     * <p>
+     * For additional info
+     * see <a href="https://www.elastic.co/guide/en/elasticsearch/reference/current/ml-put-calendar.html">
```
Comment (davidkyle) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/MachineLearningClient.java]: Wrong link and not a calendar

File: client/rest-high-level/src/main/java/org/elasticsearch/client/MachineLearningClient.java [hunk 2]
```diff
@@ -1877,4 +1882,129 @@ public void setUpgradeModeAsync(SetUpgradeModeRequest request, RequestOptions op
...
+                                                               RequestOptions options) throws IOException {
+        return restHighLevelClient.performRequestAndParseEntity(request,
+            MLRequestConverters::putDataFrameAnalytics,
+            options,
+            PutDataFrameAnalyticsResponse::fromXContent,
+            Collections.emptySet());
+    }
+
+    /**
+     * Create a new machine learning calendar, notifies listener with the created calendar
```
Comment (davidkyle) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/MachineLearningClient.java]: Calendar again. Same for get and delete below

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ValidationException.java [hunk 3]
```diff
@@ -27,6 +27,13 @@
  * Encapsulates an accumulation of validation errors
  */
 public class ValidationException extends IllegalArgumentException {
+
+    public static ValidationException withError(String error) {
```
Comment (davidkyle) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/ValidationException.java]: 👍

File: client/rest-high-level/src/test/java/org/elasticsearch/client/MachineLearningIT.java [hunk 4]
```diff
@@ -1230,6 +1242,147 @@ public void testDeleteCalendarEvent() throws IOException {
         assertThat(remainingIds, not(hasItem(deletedEvent)));
     }
 
+    public void testPutDataFrameAnalyticsConfig() throws Exception {
```
Comment (davidkyle) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/MachineLearningIT.java]: I'm surprised we don't need any test teardown in the `@After    public void cleanUp()` method. Multiple tests create a config with the id 'test-config'

File: client/rest-high-level/src/test/java/org/elasticsearch/client/ESRestHighLevelClientTestCase.java [hunk 5]
```diff
@@ -131,7 +134,12 @@ protected static RestHighLevelClient highLevelClient() {
 
     private static class HighLevelClient extends RestHighLevelClient {
         private HighLevelClient(RestClient restClient) {
-            super(restClient, (client) -> {}, new SearchModule(Settings.EMPTY, false, Collections.emptyList()).getNamedXContents());
+            super(
+                restClient,
+                (client) -> {},
+                Stream.concat(
+                    new SearchModule(Settings.EMPTY, false, Collections.emptyList()).getNamedXContents().stream(),
+                    new MlDataFrameAnalysisNamedXContentProvider().getNamedXContentParsers().stream()).collect(toList()));
```
Comment (davidkyle) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/ESRestHighLevelClientTestCase.java]: It's a bit untidy that this has to be done here. `EsIntegTestCase` has the more extensible approach of using a virtual method to gather the NamedXContentRegistry entries. What do yo think of doing something similar?

```
    @Override
    protected NamedXContentRegistry xContentRegistry() 
```

**Review by davidkyle — COMMENTED** [review]
Looks great. Left some minor comments

File: client/rest-high-level/src/test/java/org/elasticsearch/client/MachineLearningIT.java (hunk 4 above)
Comment (przemekwitek) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/MachineLearningIT.java]: Done, good catch.
I don't know why I didn't hit me, maybe I was lucky with test execution order on my machine?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/MachineLearningClient.java (hunk 1 above)
Comment (davidkyle) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/MachineLearningClient.java]: The documentation for these endpoints isn't up yet. See https://github.com/elastic/elasticsearch/blob/feature-ml-data-frame-analytics/x-pack/plugin/src/test/resources/rest-api-spec/api/ml.get_data_frame_analytics.json which has no 'documentation' field. 

Make the links TODOs for now and we can revisit this

Comment (przemekwitek) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/MachineLearningClient.java]: Done.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/MachineLearningClient.java (hunk 2 above)
Comment (przemekwitek) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/MachineLearningClient.java]: Done.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/ESRestHighLevelClientTestCase.java (hunk 5 above)
Comment (przemekwitek) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/ESRestHighLevelClientTestCase.java]: Virtual method approach doesn't work for integration test classes (I don't why).
But I was able to solve this by adding my XContentProvider to client/rest-high-level/src/main/resources/META-INF/services/org.elasticsearch.plugins.spi.NamedXContentProvider file.

Comment (davidkyle) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/ESRestHighLevelClientTestCase.java]: Neat I hadn't thought of that.

**Review by davidkyle — APPROVED** [review]
LGTM. The docs will need doing but we can save that for a later change

**Pull request merged**